app.go

```diff
@@ -930,22 +930,24 @@ func (a *Application) ReadOnlyState() State {
 	)
 }
 
+// ReadOnlyStateAt returns the app state as it was at a past block height, or an error if the state
+// couldn't be loaded for that block (likely due to pruning).
 func (a *Application) ReadOnlyStateAt(version int64) (State, error) {
-	readOnlySnapshot, err := a.Store.GetSnapshotAt(version)
+	snapshot, err := a.Store.GetSnapshotAt(version)
 	if err != nil {
 		return nil, err
 	}
-	block, err := a.BlockStore.GetBlockByHeight(&version)
+	blockInfo, err := a.BlockStore.GetBlockByHeight(&version)
 	if err != nil {
 		return nil, err
 	}
 	// TODO: the store snapshot should be created atomically, otherwise the block header might
 	//       not match the state... need to figure out why this hasn't spectacularly failed already
 	return NewStoreStateSnapshot(
 		nil,
-		readOnlySnapshot,
-		abciHeaderFromBlockHeader(block.Block.Header),
-		nil, // TODO: last block hash!
+		snapshot,
+		abciHeaderFromBlockHeader(blockInfo.Block.Header),
+		blockInfo.Block.Hash(),
```

Comment:
Block.Hash() always returns an empty hash. Not sure why but this makes `ReadOnlyStateAt` snapshot not work properly. I think we have to set it to nil for now.